README.md
# RPostgreSQL
An R Interface to PostgreSQL.


Releases go to CRAN at
https://CRAN.R-project.org/package=RPostgreSQL

Check status of the release in various platforms can be found at
https://cran.r-project.org/web/checks/check_results_RPostgreSQL.html

Tests automatically run at Travis ci can be found
https://travis-ci.org/tomoakin/RPostgreSQL


inst/NEWS
Version 0.5-2

    o   Support dbBegin() API

version 0.5-1 --2014-02-08

    o   Remove segmentationfault when NULL is passed as host argument.

    o   Conversion routine for raw in R and bytea in PostgreSQL

    o   Use accompanied libpq when either of libpq and header is not found on darwin


Version 0.5 -- 2013-10-12

    o   postgresqlWriteTable: send column names at COPY

    o   System library header search path for OS X 10.9 (Mavericks)

    o   Memory leak was reduced
    o   Close the socket on failure

Version 0.4 -- 2013-03-27

    o   Initial implementation of prepared statement

    o   Use system libpq library when available on OS X (darwin)

    o   Force ISO datestyle for PostgreSQL-R communication by default

    o   Time zone aware Timestamp POSIXct conversion

Version 0.3-3 -- 2012-10-05

    o   Bugfix on dbColumnInfo and others reproducible by gctorture(TRUE)

    o   Do not implicitly make new connections, which interfere with transaction.

    o   Change the notation of the LICENSE

Version 0.3-2 -- 2012-01-10

    o   Adapt to the new windows toolchain.

    o   Do not make libpq.dll on windows but just make libpq.a and link it,
        so that libpq.dll need not copied anymore.

Version 0.3-1 -- 2011-12-31

    o   Remove GNU make dependency.

Version 0.3-0 -- 2011-12-29

    o   The bundled libpq source codes are used under darwin as well.

    o   More compatibility to other database drivers under DBI and sqldf.
        dbBeginTransaction, safe.write, make.db.names are removed.
        dbBuildTableDefinition is renamed to postgresqlBuildTableDefinition.
        isIdCurrent is renamed to isPostgresqlIdCurrent.
        row.names is now integer.

Version 0.2-1 -- 2011-11-16

    o   libpq source codes are bundled for Windows.
	In other envirionment, this code is not used and existence of libpq as 
	specified by PG_HOME is still required.

Version 0.2-0 -- 2011-10-04

    o   Error check for dbWriteTable

    o   dbWriteTable does not write a temporary file.
        Control characters in strings are properly escaped for dbWriteTable.

    o   Remove warning for enum and arrays etc.  The value are still transferred
        as string and not converted to arrays etc.

    o   use c('schema', 'table') to specify schema

    o   Integer is mapped to integer in postgresql rather than bigint.

Version 0.1-7 -- 2010-10-17

    o   Several potential buffer overruns were fixed

    o   dbWriteTable now writes a data.frame to database through a network
        connection rather than a temporary file. Note that row_names may be
        changed in future releases.  Also, passing in filenames instead of
        data.frame is not supported at this time. 

    o   When no host is specified, a connection to the PostgreSQL server 
        is made via UNIX domain socket (just like psql does)

    o   Table and column names are case sensitive, and identifiers are escaped
        or quoted appropriately, so that any form of table/column names can be
        created, searched, or removed, including upper-, lower- and mixed-case.

    o   nullOk in dbColumnInfo has a return value of NA when the column does
        not correspond to a column in the table. The utility of nullOk is
        doubtful but not removed at this time.

    o   Correct Windows getpid() declaration (with thanks to Brian D. Ripley)

    o   A call of as.POSIXct() with a time format string wrongly passed to TZ
        has been corrected; this should help with intra-day timestamps (with
        thanks to Steve Eick)

    o   Usage of tmpdir has been improved on similarly to Linux (with thanks
        to Robert McGehee)

Version 0.1-6 -- 2009-10-19

    o   Added missing paste() call to dbGetQuery() for dbListTables

Version 0.1-5 -- 2009-10-13

    o   Four issues reported at the Issue tracker at the Google Code site
        are addressed, two more are feature requests and one cannot be
        replicated

    o   A number of other small fixes and enhancements to code and
        documentation as detailed in the ChangeLog file

Version 0.1-4 -- 2009-01-26
   
    o   Fix to one of the documentation files

Version 0.1-3 -- 2008-12-12

    o   Fixed a memory leak detected by Valgrind with thanks to Jeff Horner
        who applied a similar fix to RMySQL

Version 0.1-2 -- 2008-11-02

    o   Some fixes to the regression tests and configuration

Version 0.1-1 -- 2008-10-28

    o	DESCRIPTION: Correct Url: field by adding http:// parts. Thanks to
        Gabor for the hint.

Version 0.1-0 -- 2008-10-21

    o	First Release

    o   This 'RPostgreSQL' package was developed as a part of Google Summer of
     	Code 2008 program.

    o   Its implements all the features in the DBI (and a few extra like those
        related to transaction management)


R/PostgreSQL.R
## PostgreSQL.R
## $Id$

## This package was developed as a part of Summer of Code program organized by Google.
## Thanks to David A. James & Saikat DebRoy, the authors of RMySQL package.
## Code from RMySQL package was reused with the permission from the authors.
## Also Thanks to my GSoC mentor Dirk Eddelbuettel for helping me in the development.

##
## Constants
##

##.PostgreSQLRCS <- "$Id: PostgreSQL.R,v 0.1 2008/06/10 14:00:00$"
.PostgreSQLPkgName <- "RPostgreSQL"
.PostgreSQLVersion <- "0.5-2"       ##package.description(.PostgreSQLPkgName, fields = "Version")
.PostgreSQL.NA.string <- "\\N"      ## on input, PostgreSQL interprets \N as NULL (NA)

setOldClass("data.frame")      ## to appease setMethod's signature warnings...

##
## Class: DBIObject
##
setClass("PostgreSQLObject", representation("DBIObject", "dbObjectId", "VIRTUAL"))

##
## Class: dbDriver
##

PostgreSQL <- function(max.con=16, fetch.default.rec = 500, force.reload=FALSE) {
    postgresqlInitDriver(max.con = max.con, fetch.default.rec = fetch.default.rec,
                         force.reload = force.reload)
}

##
## Class: DBIDriver
##
setClass("PostgreSQLDriver", representation("DBIDriver", "PostgreSQLObject"))

## coerce (extract) any PostgreSQLObject into a PostgreSQLDriver
setAs("PostgreSQLObject", "PostgreSQLDriver",
      def = function(from) new("PostgreSQLDriver", Id = as(from, "integer")[1:2])
      )

setMethod("dbUnloadDriver", "PostgreSQLDriver",
          def = function(drv, ...) postgresqlCloseDriver(drv, ...),
          valueClass = "logical"
          )

setMethod("dbGetInfo", "PostgreSQLDriver",
          def = function(dbObj, ...) postgresqlDriverInfo(dbObj, ...)
          )

setMethod("dbListConnections", "PostgreSQLDriver",
          def = function(drv, ...) dbGetInfo(drv, "connectionIds")[[1]]
          )

setMethod("summary", "PostgreSQLDriver",
          def = function(object, ...) postgresqlDescribeDriver(object, ...)
          )

##
## Class: DBIConnection
##
setClass("PostgreSQLConnection", representation("DBIConnection", "PostgreSQLObject"))

setMethod("dbConnect", "PostgreSQLDriver",
          def = function(drv, ...) postgresqlNewConnection(drv, ...),
          valueClass = "PostgreSQLConnection"
          )

setMethod("dbConnect", "character",
          def = function(drv, ...) postgresqlNewConnection(dbDriver(drv), ...),
          valueClass = "PostgreSQLConnection"
          )

## clone a connection
setMethod("dbConnect", "PostgreSQLConnection",
          def = function(drv, ...) postgresqlCloneConnection(drv, ...),
          valueClass = "PostgreSQLConnection"
          )

setMethod("dbDisconnect", "PostgreSQLConnection",
          def = function(conn, ...) postgresqlCloseConnection(conn, ...),
          valueClass = "logical"
          )

setGeneric("dbEscapeStrings", def = function(conn, string, ...) standardGeneric("dbEscapeStrings"))
setMethod("dbEscapeStrings",
          signature(conn="PostgreSQLConnection", string="character"),
          def = function(conn, string, ...) postgresqlEscapeStrings(conn, string, ...),
          valueClass = "character"
          )

setMethod("dbSendQuery",
          signature(conn = "PostgreSQLConnection", statement = "character"),
          def = function(conn, statement,...) postgresqlExecStatement(conn, statement,...),
          valueClass = "PostgreSQLResult"
          )


setMethod("dbGetQuery",
          signature(conn = "PostgreSQLConnection", statement = "character"),
          def = function(conn, statement, ...) postgresqlQuickSQL(conn, statement, ...)
          )

setMethod("dbGetException", "PostgreSQLConnection",
          def = function(conn, ...){
              if(!isPostgresqlIdCurrent(conn))
                  stop(paste("expired", class(conn)))
              .Call("RS_PostgreSQL_getException", as(conn, "integer"),
                    PACKAGE = .PostgreSQLPkgName)
          },
          valueClass = "list"
          )

setMethod("dbGetInfo", "PostgreSQLConnection",
          def = function(dbObj, ...) postgresqlConnectionInfo(dbObj, ...)
          )

setMethod("dbListResults", "PostgreSQLConnection",
          def = function(conn, ...) dbGetInfo(conn, "rsId")[[1]]
          )

setMethod("summary", "PostgreSQLConnection",
          def = function(object, ...) postgresqlDescribeConnection(object, ...)
          )

## convenience methods
setMethod("dbListTables", "PostgreSQLConnection",
          def = function(conn, ...){
              out <- dbGetQuery(conn,
                                paste("select tablename from pg_tables where schemaname !='information_schema'",
                                      "and schemaname !='pg_catalog'", ...))
              if (is.null(out) || nrow(out) == 0)
                  out <- character(0)
              else
                  out <- out[, 1]
              out
          },
          valueClass = "character"
          )

setMethod("dbReadTable", signature(conn="PostgreSQLConnection", name="character"),
          def = function(conn, name, ...) postgresqlReadTable(conn, name, ...),
          valueClass = "data.frame"
          )

setMethod("dbWriteTable",
          signature(conn="PostgreSQLConnection", name="character", value="data.frame"),
          def = function(conn, name, value, ...){
              postgresqlWriteTable(conn, name, value, ...)
          },
          valueClass = "logical"
          )

## write table from filename (TODO: connections)
setMethod("dbWriteTable",
          signature(conn="PostgreSQLConnection", name="character", value="character"),
          def = function(conn, name, value, ...){
              postgresqlImportFile(conn, name, value, ...)
          },
          valueClass = "logical"
          )

setMethod("dbExistsTable",
          signature(conn="PostgreSQLConnection", name="character"),
          def = function(conn, name, ...){
              qlength <- length(name)
              if(qlength == 1){
              currentschema <- dbGetQuery(conn, "SELECT current_schema()")
              res <- dbGetQuery(conn,
                  paste("select tablename from pg_tables where ",
                  "schemaname !='information_schema' and schemaname !='pg_catalog' ",
                  "and schemaname='", postgresqlEscapeStrings(conn, currentschema[[1]]), "' ",
                  "and tablename='", postgresqlEscapeStrings(conn, name), "'", sep=""))
              }
              else{
                  if(qlength == 2){
                  res <- dbGetQuery(conn,
                      paste("select tablename from pg_tables where ",
                      "schemaname !='information_schema' and schemaname !='pg_catalog' ",
                      "and schemaname='", postgresqlEscapeStrings(conn, name[1]), "' ",
                      "and tablename='", postgresqlEscapeStrings(conn, name[2]), "'", sep=""))
                  }
              }
              return(as.logical(dim(res)[1]))
          },
          valueClass = "logical"
          )

setMethod("dbRemoveTable",
          signature(conn="PostgreSQLConnection", name="character"),
          def = function(conn, name, ...){
              if(dbExistsTable(conn, name)){
                  rc <- try(dbGetQuery(conn, paste("DROP TABLE", postgresqlTableRef(name))))
                  !inherits(rc, ErrorClass)
              }
              else FALSE
          },
          valueClass = "logical"
          )

setMethod("dbBegin",
          signature(conn="PostgreSQLConnection"),
          def = function(conn,  ...){
              rc <- try(dbGetQuery(conn, "BEGIN"))
              !inherits(rc, ErrorClass)
          },
          valueClass = "logical"
          )
setMethod("dbCommit",
          signature(conn="PostgreSQLConnection"),
          def = function(conn,  ...){
              rc <- try(dbGetQuery(conn, "COMMIT"))
              !inherits(rc, ErrorClass)
          },
          valueClass = "logical"
          )
setMethod("dbRollback",
          signature(conn="PostgreSQLConnection"),
          def = function(conn,  ...){
              rc <- try(dbGetQuery(conn, "ROLLBACK"))
              !inherits(rc, ErrorClass)
          },
          valueClass = "logical"
          )

## return field names (no metadata)
setMethod("dbListFields",
          signature(conn="PostgreSQLConnection", name="character"),
          def = function(conn, name, ...){
              qlength <- length(name)
              if(qlength == 1){
              currentschema <- dbGetQuery(conn, "SELECT current_schema()")
              flds <- dbGetQuery(conn,
                  paste("select a.attname from pg_attribute a, pg_class c, pg_tables t, pg_namespace nsp",
                  " where a.attrelid = c.oid and c.relname = tablename and c.relnamespace = nsp.oid and a.attnum > 0 and ",
                  "nspname = current_schema() and schemaname = nspname and ",
                  "tablename = '", postgresqlEscapeStrings(conn, name), "'", sep=""))[,1]
              }
              else{
                  if(qlength == 2){
                  flds <- dbGetQuery(conn,
                      paste("select a.attname from pg_attribute a, pg_class c, pg_tables t, pg_namespace nsp",
                      " where a.attrelid = c.oid and c.relname = t.tablename and c.relnamespace = nsp.oid and a.attnum > 0 and ",
                      "nspname = schemaname ",
                      "and schemaname = '", postgresqlEscapeStrings(conn, name[1]), "' ",
                      "and tablename = '", postgresqlEscapeStrings(conn, name[2]), "'", sep=""))[,1]
                  }
              }

              if(length(flds)==0)
                  flds <- character()
              flds
          },
          valueClass = "character"
          )


setMethod("dbCallProc", "PostgreSQLConnection",
          def = function(conn, ...) .NotYetImplemented()
          )

setMethod("dbCommit", "PostgreSQLConnection",
          def = function(conn, ...) postgresqlTransactionStatement(conn, "COMMIT")
          )

setMethod("dbRollback", "PostgreSQLConnection",
          def = function(conn, ...) {
              rsList <- dbListResults(conn)
              if (length(rsList))
                  dbClearResult(rsList[[1]])
              postgresqlTransactionStatement(conn, "ROLLBACK")
          }
          )


##
## Class: DBIResult
##
setClass("PostgreSQLResult", representation("DBIResult", "PostgreSQLObject"))

setAs("PostgreSQLResult", "PostgreSQLConnection",
      def = function(from) new("PostgreSQLConnection", Id = as(from, "integer")[1:3])
      )
setAs("PostgreSQLResult", "PostgreSQLDriver",
      def = function(from) new("PostgreSQLDriver", Id = as(from, "integer")[1:2])
      )

setMethod("dbClearResult", "PostgreSQLResult",
          def = function(res, ...) postgresqlCloseResult(res, ...),
          valueClass = "logical"
          )

setMethod("fetch", signature(res="PostgreSQLResult", n="numeric"),
          def = function(res, n, ...){
              out <- postgresqlFetch(res, n, ...)
              if(is.null(out))
                  out <- data.frame(out)
              out
          },
          valueClass = "data.frame"
          )

setMethod("fetch",
          signature(res="PostgreSQLResult", n="missing"),
          def = function(res, n, ...){
              out <-  postgresqlFetch(res, n=0, ...)
              if(is.null(out))
                  out <- data.frame(out)
              out
          },
          valueClass = "data.frame"
          )

setMethod("dbGetInfo", "PostgreSQLResult",
          def = function(dbObj, ...) postgresqlResultInfo(dbObj, ...),
          valueClass = "list"
          )

setMethod("dbGetStatement", "PostgreSQLResult",
          def = function(res, ...){
              st <-  dbGetInfo(res, "statement")[[1]]
              if(is.null(st))
                  st <- character()
              st
          },
          valueClass = "character"
          )

setMethod("dbListFields",
          signature(conn="PostgreSQLResult", name="missing"),
          def = function(conn, name, ...){
              flds <- dbGetInfo(conn, "fields")$fields$name
              if(is.null(flds))
                  flds <- character()
              flds
          },
          valueClass = "character"
          )

setMethod("dbColumnInfo", "PostgreSQLResult",
          def = function(res, ...) postgresqlDescribeFields(res, ...),
          valueClass = "data.frame"
          )

setMethod("dbGetRowsAffected", "PostgreSQLResult",
          def = function(res, ...) dbGetInfo(res, "rowsAffected")[[1]],
          valueClass = "numeric"
          )

setMethod("dbGetRowCount", "PostgreSQLResult",
          def = function(res, ...) dbGetInfo(res, "rowCount")[[1]],
          valueClass = "numeric"
          )

setMethod("dbHasCompleted", "PostgreSQLResult",
          def = function(res, ...) dbGetInfo(res, "completed")[[1]] == 1,
          valueClass = "logical"
          )

setMethod("dbGetException", "PostgreSQLResult",
          def = function(conn, ...){
              id <- as(conn, "integer")[1:2]
              .Call("RS_PostgreSQL_getException", id, PACKAGE = .PostgreSQLPkgName)
          },
          valueClass = "list"    ## TODO: should be a DBIException?
          )

setMethod("summary", "PostgreSQLResult",
          def = function(object, ...) postgresqlDescribeResult(object, ...)
          )



setMethod("dbDataType",
          signature(dbObj = "PostgreSQLObject", obj = "ANY"),
          def = function(dbObj, obj, ...) postgresqlDataType(obj, ...),
          valueClass = "character"
          )


## MODIFIED : -- sameer
setMethod("make.db.names",
          signature(dbObj="PostgreSQLObject", snames = "character"),
          def = function(dbObj, snames,keywords,unique, allow.keywords,...){
              postgresqlQuoteId(snames)
          },
          valueClass = "character"
          )

setMethod("SQLKeywords", "PostgreSQLObject",
          def = function(dbObj, ...) .PostgreSQLKeywords,
          valueClass = "character"
          )

setMethod("isSQLKeyword",
          signature(dbObj="PostgreSQLObject", name="character"),
          def = function(dbObj, name,keywords,case, ...){
              isSQLKeyword.default(name, keywords = .PostgreSQLKeywords)
          },
          valueClass = "character"
          )
## extension to the DBI 0.1-4
setGeneric("dbApply", def = function(res, ...) standardGeneric("dbApply"))
setMethod("dbApply", "PostgreSQLResult",
          def = function(res, ...)  postgresqlDBApply(res, ...),
          )


